### PR TITLE
kafka_topic_enforcer: indicate replication factor config drift

### PR DESCRIPTION
**Pull request by mattbonnell**
#18

### CODE REVIEW

**Comment by shrijeet-tesla** [comment]
Let's add some logging when an unsupported drift is detected. Also, what happens when we have 

a/ a safe drift, along with an unsupported drift (ex: change partitions + change replication)
b/ an un-safe drift along with an un-supported drift (ex: 3x partitions + change replication)

**Comment by mattbonnell** [comment]
> Let's add some logging when an unsupported drift is detected. 

At which point in the code are you thinking? Ex currently it would get logged [here](https://github.com/teslamotors/kafka-helmsman/blob/b49eb2e95129e6c7f0123c53548c0b995968615b/kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/TopicEnforcer.java#L93) during the TopicEnforcer run.

>Also, what happens when we have
> 
> a/ a safe drift, along with an unsupported drift (ex: change partitions + change replication)
> b/ an un-safe drift along with an un-supported drift (ex: 3x partitions + change replication)

It seems the config properties are only ever checked separately (eg [here](https://github.com/teslamotors/kafka-helmsman/blob/b49eb2e95129e6c7f0123c53548c0b995968615b/kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/TopicEnforcer.java#L158-L173)), so I would imagine we get a log message and metric emitted for each detected drift. As far as I can tell from git-grepping, the ANY drift type isn't used anywhere. But according to the current implementation of ConfigDrift.check(), if ANY drift type is passed in, the result will either be NO_DRIFT if actual config == desired config or SAFETY_UNKNOWN_DRIFT otherwise (no matter what the drift(s) are).

**Comment by shrijeet-tesla** [comment]
> At which point in the code are you thinking? Ex currently it would get logged here during the TopicEnforcer run.

Good point. I was thinking we can add a separate warning to highlight that we have some un-supported drifts. But if we do that, we should do that for unsafe drifts too.

> It seems the config properties are only ever checked separately (eg here), so I would imagine we get a log message and metric emitted for each detected drift.

So in my example (a), metrics & logging aside, would we get replicas increased but replication ignored? Can you add a test case.

**Comment by mattbonnell** [comment]
added test cases to the TopicEnforcer suite, that suite was failing to begin with though (fails on master) so my next step is to debug that to ensure new changes are working correctly there

**Comment by mattbonnell** [comment]
all tests now passing.
As per case 1 in https://stackoverflow.com/a/41265267, the flag `--add-opens java.base/java.lang=ALL-UNNAMED` needs to be passed to the JVM in order for mockito's reflection to work, thus `bazel test` must be supplied the following argument: `--jvmopt="--add-opens java.base/java.lang=ALL-UNNAMED"`

File: kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/ConfigDrift.java [hunk 1]
```diff
@@ -97,6 +101,9 @@ public Result check(ConfiguredTopic desired, ConfiguredTopic actual, Type type)
           }
         }
 
+      case REPLICATION_FACTOR:
+	return desired.getReplicationFactor() == actual.getReplicationFactor() ? Result.NO_DRIFT: Result.UNSUPPORTED_DRIFT;
```
Comment (shrijeet-tesla) [inline kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/ConfigDrift.java:105]: this has odd indentation. Is your IDE setup with correct code style?

File: kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/TopicEnforcer.java [hunk 2]
```diff
@@ -169,5 +187,8 @@ public void stats() {
     topicConfigDrift
         .labels("unsafe")
         .set(topicsWithConfigDrift(Type.TOPIC_CONFIG, Result.UNSAFE_DRIFT).size());
+    replicationFactorDrift
+	.labels("unsupported")
```
Comment (shrijeet-tesla) [inline kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/TopicEnforcer.java:191]: odd formatting

File: kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/ConfigDrift.java (hunk 1 above)
Comment (mattbonnell) [reply kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/ConfigDrift.java]: it wasn't at the time. I've got it setup with autoformat and two-space indentation now. I notice though that it seems my autoformatter is making some different decisions with regards to breaking up lines than in the rest of the code base. do you have a particular style configuration which you use?

File: kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/TopicEnforcer.java [hunk 3]
```diff
@@ -89,11 +97,21 @@ public TopicEnforcer(
                 t -> {
                   Result result = configDrift.check(t, existing.get(t.getName()), type);
                   if (logResults && !Result.NO_DRIFT.equals(result)) {
-                    LOG.info(
-                        "Found {} for topic {}, in {} drift detection mode",
-                        result,
-                        t.getName(),
-                        type);
+		    String msgFmt = "Found {} for topic {}, in {} drift detection mode"; 
+		    if (result.equals(Result.SAFE_DRIFT)) {
```
[the remark this answers is not in the document]
Comment (mattbonnell) [reply kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/TopicEnforcer.java]: fixed

File: kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/TopicEnforcer.java [hunk 4]
```diff
@@ -40,6 +40,14 @@
           .labelNames("type")
           .register();
 
+  private static final Gauge replicationFactorDrift =
+      Gauge.build()
+          .name("kafka_topic_enforcer_replication_factor_drift_topics")
+          .help("Count of topics which have drifted from their desired replication factor.")
+          .labelNames("type")
```
[the remark this answers is not in the document]
Comment (mattbonnell) [reply kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/TopicEnforcer.java]: fixed.

File: kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/TopicEnforcer.java (hunk 2 above)
Comment (mattbonnell) [reply kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/TopicEnforcer.java]: fixed.

File: kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/ConfigDrift.java [hunk 5]
```diff
@@ -15,27 +15,30 @@
  */
 public class ConfigDrift {
 
-  // permissible % increase threshold doesnt apply if the increased amount is below low water mark
+  // permissible % increase threshold doesnt apply if the increased amount is
+  // below low water mark
```
Comment (jyates) [inline kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/ConfigDrift.java:19]: it makes reviewing much easier if you don't change unrelated lines like these. In fact, changing these lines looks wrong - we should be supporting a 120 wide line

File: kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/ConfigDrift.java [hunk 6]
```diff
@@ -15,27 +15,30 @@
...
+  public ConfigDrift(int partitionCountLowWaterMark, int partitionCountHighWaterMark, float partitionIncreaseThreshold,
+      Set<String> unsafeConfigProperties) {
     checkArgument(partitionIncreaseThreshold > 0 && partitionIncreaseThreshold <= 1);
     checkArgument(partitionCountLowWaterMark > 0 && partitionCountHighWaterMark > 0);
     this.partitionCountLowWaterMark = partitionCountLowWaterMark;
     this.partitionCountHighWaterMark = partitionCountHighWaterMark;
     this.partitionIncreaseThreshold = partitionIncreaseThreshold;
-    this.unsafeConfigProperties = checkNotNull(unsafeConfigProperties);;
+    this.unsafeConfigProperties = checkNotNull(unsafeConfigProperties);
+    ;
```
Comment (jyates) [inline kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/ConfigDrift.java:37]: nit: remove the extra semi-colon?

File: kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/ConfigDrift.java [hunk 7]
```diff
@@ -55,15 +60,20 @@ public ConfigDrift() {
...
+    UNSUPPORTED_DRIFT
   }
 
   /**
    * Check if there is a config drift between two topic configuration states.
    *
-   * @param desired the desired {@link ConfiguredTopic}
-   * @param actual an actual {@link ConfiguredTopic}
-   * @param type the checking mode, see {@link Type}
+   * @param desired
```
Comment (jyates) [inline kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/ConfigDrift.java:71]: can you roll back these changes? looks like your IDE formatter does not match the existing format.

File: kafka_topic_enforcer/src/test/java/com/tesla/data/topic/enforcer/TopicEnforcerTest.java [hunk 8]
```diff
@@ -48,16 +48,18 @@ public void testPartitionIncrease() {
...
     Map<String, ConfiguredTopic> existing = new HashMap<String, ConfiguredTopic>() {
       {
         // cluster topic has less replicas than configured but same partitions, we do not care!
         put("a", new ConfiguredTopic("a", 1, (short) 1, Collections.emptyMap()));
         // cluster topic has less partitions than configured, we care!
         put("b", new ConfiguredTopic("b", 1, (short) 1, Collections.emptyMap()));
+        // cluster topic has less partitions than configured and less replicas, we care (about the partitions)!
+        put("c", new ConfiguredTopic("c", 1, (short) 1, Collections.emptyMap()));
         // cluster has an extra topic that is not configured to be present, we do not care!
-        put("d", new ConfiguredTopic("b", 1, (short) 1, Collections.emptyMap()));
```
Comment (jyates) [inline kafka_topic_enforcer/src/test/java/com/tesla/data/topic/enforcer/TopicEnforcerTest.java:60]: what is wrong with name `d`?

File: kafka_topic_enforcer/src/test/java/com/tesla/data/topic/enforcer/ConfigDriftTest.java [hunk 9]
```diff
@@ -72,6 +73,18 @@ public void testUnSafeTopicProperties() {
...
 
+  @Test
+  public void testReplicationFactorDrift() {
+	  ConfiguredTopic desired = new ConfiguredTopic("a", 10, (short) 2, Collections.emptyMap());
+	  Assert.assertEquals(UNSUPPORTED_DRIFT, drift.check(desired, actual, ConfigDrift.Type.REPLICATION_FACTOR));
+  }
+
+  @Test
+  public void testReplicationFactorNoDrift() {
+	  ConfiguredTopic desired = new ConfiguredTopic("a", 10, (short) 1, Collections.emptyMap());
```
Comment (jyates) [inline kafka_topic_enforcer/src/test/java/com/tesla/data/topic/enforcer/ConfigDriftTest.java:84]: should be 2 spaces per indent

File: kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/TopicEnforcer.java [hunk 10]
```diff
@@ -157,17 +143,12 @@ protected void delete(List<ConfiguredTopic> toDelete) {
   @Override
   public void stats() {
     super.stats();
-    partitionCountDrift
```
Comment (jyates) [inline kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/TopicEnforcer.java:160]: this was formatted for ease of readability.

**Review by jyates — CHANGES_REQUESTED** [review]
All the formatting changes make it very hard to get to the meat of the change. From what i can tell, that does look OK, but hard to be sure

File: kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/ConfigDrift.java (hunk 5 above)
Comment (mattbonnell) [reply kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/ConfigDrift.java:19]: yeah this is my bad, didn't have my IDE properly configured style-wise and the auto-formatter wreaked a bit of havoc. correcting this now, will get the unrelated formatting changes out.

File: kafka_topic_enforcer/src/test/java/com/tesla/data/topic/enforcer/TopicEnforcerTest.java (hunk 8 above)
Comment (mattbonnell) [reply kafka_topic_enforcer/src/test/java/com/tesla/data/topic/enforcer/TopicEnforcerTest.java:60]: good question, not sure how this change came about, will revert

**Comment by mattbonnell** [comment]
Removed all the formatting changes

File: kafka_topic_enforcer/src/test/java/com/tesla/data/topic/enforcer/TopicEnforcerTest.java [hunk 11]
```diff
@@ -95,7 +100,7 @@ public void testAlterConfiguration() {
     enforcer.alterConfiguration();
 
     // topic 'a'
-    List<ConfiguredTopic> expected = Collections.singletonList(configured.get(0));
+    List<ConfiguredTopic> expected = Arrays.asList(configured.get(0), configured.get(3));
```
Comment (shrijeet-tesla) [inline kafka_topic_enforcer/src/test/java/com/tesla/data/topic/enforcer/TopicEnforcerTest.java:103]: is this change needed? If yes, can you fix the comment, looks like its out of date

File: kafka_topic_enforcer/src/test/java/com/tesla/data/topic/enforcer/TopicEnforcerTest.java [hunk 12]
```diff
@@ -66,7 +68,7 @@ public void testPartitionIncrease() {
     enforcer.increasePartitions();
 
     // topic 'b'
-    List<ConfiguredTopic> expected = Collections.singletonList(configured.get(1));
+    List<ConfiguredTopic> expected = Arrays.asList(configured.get(1), configured.get(2));
```
Comment (shrijeet-tesla) [inline kafka_topic_enforcer/src/test/java/com/tesla/data/topic/enforcer/TopicEnforcerTest.java:71]: is comment in-sync with latest change?

**Review by shrijeet-tesla — APPROVED** [review]
Fix comments in test and then good to 🚢

File: kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/TopicEnforcer.java [hunk 13]
```diff
@@ -89,11 +95,12 @@ public TopicEnforcer(
                 t -> {
                   Result result = configDrift.check(t, existing.get(t.getName()), type);
                   if (logResults && !Result.NO_DRIFT.equals(result)) {
-                    LOG.info(
-                        "Found {} for topic {}, in {} drift detection mode",
-                        result,
-                        t.getName(),
-                        type);
+                    String msgFmt = "Found {} for topic {}, in {} drift detection mode";
```
Comment (jyates) [inline kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/TopicEnforcer.java:98]: have you tried a switch statement here? Might be cleaner since we are dealing with at least 3 different result types now

**Review by jyates — APPROVED** [review]
lgtm

File: kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/TopicEnforcer.java (hunk 13 above)
Comment (mattbonnell) [reply kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/TopicEnforcer.java:98]: done

File: kafka_topic_enforcer/src/test/java/com/tesla/data/topic/enforcer/TopicEnforcerTest.java (hunk 12 above)
Comment (mattbonnell) [reply kafka_topic_enforcer/src/test/java/com/tesla/data/topic/enforcer/TopicEnforcerTest.java:71]: addressed

File: kafka_topic_enforcer/src/test/java/com/tesla/data/topic/enforcer/TopicEnforcerTest.java (hunk 11 above)
Comment (mattbonnell) [reply kafka_topic_enforcer/src/test/java/com/tesla/data/topic/enforcer/TopicEnforcerTest.java:103]: updated that line because I added another test case there, another topic which should be changing (topic "d"). updated the comments to reflect it

File: kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/TopicEnforcer.java [hunk 14]
```diff
@@ -88,12 +94,18 @@ public TopicEnforcer(
...
-                        type);
+                  if (logResults) {
+                    String msgFmt = "Found {} for topic {}, in {} drift detection mode";
+                    switch(result) {
+                      case NO_DRIFT:
+                        break;
+                      case SAFE_DRIFT:
+                        LOG.info(msgFmt, result, t.getName(), type);
+                        break;
+                      default:
```
Comment (jyates) [inline kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/TopicEnforcer.java:105]: would be good to explicitly call out the UNSAFE enum here, for readability sake

Comment (mattbonnell) [reply kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/TopicEnforcer.java]: refactored this, let me know what you think
